app/controllers/documents_controller.rb

```diff
@@ -254,7 +253,20 @@ def update
 
   protected
   def can_use_edit_token
-    if session[:lms_authenticated_user] || @organization[:enable_anonymous_actions] || has_role("designer")
+    lms_connection_information
```

Comment:
Move lms_course_id to an argument, pass in when relevant
Clean up these checks, should do the least amount of work first (simple boolean checks), then do the canvas check last
Move lms session checks to helper method
Move canvas code to seperate method as well